runtime/src/bank_forks.rs

```diff
@@ -260,14 +277,23 @@ impl BankForks {
     }
 
     fn prune_non_root(&mut self, root: Slot, highest_confirmed_root: Option<Slot>) {
-        let descendants = self.descendants();
-        self.banks.retain(|slot, _| {
-            *slot == root
-                || descendants[&root].contains(slot)
-                || (*slot < root
-                    && *slot >= highest_confirmed_root.unwrap_or(root)
-                    && descendants[slot].contains(&root))
-        });
+        let highest_confirmed_root = highest_confirmed_root.unwrap_or(root);
+        let prune_slots: Vec<_> = self
+            .banks
+            .keys()
+            .copied()
+            .filter(|slot| {
+                let keep = *slot == root
+                    || self.descendants[&root].contains(slot)
+                    || (*slot < root
+                        && *slot >= highest_confirmed_root
+                        && self.descendants[slot].contains(&root));
+                !keep
+            })
+            .collect();
+        for slot in prune_slots {
+            self.remove(slot);
```

Comment:
Also of note, it seems ok if `descendants` contains keys that don't exist in the bank. In replay where this is used

This code is subtly scary because of cases like:
https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1595, which calls into
https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L638, does not assume the key exists in the progress map, but does assume it exists in the `ancestors` map here: https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L640-L645

Reply:
Added more tests. Please double check the tests to confirm it matches the scenario you are describing.

Regarding keys that don't exit in the bank, I don't think the logic there has changed compared to before. `descendants` will always include any slot referenced to in any of the `bank[i].ancestors` (including `banks[i].slot`), regardless of if it exists in the bank-forks or not.